man/lpass.1
.TH LPASS 1 "10 FEBRUARY 2025" "LPASS v1.3.0" LockPassword

.SH NAME
LockPassword - a password manager.

.SH SYNOPSIS
.B lpass
.RI < command >
.RI [< options >]

.SH DESCRIPTION
.B lpass
- a simple terminal password manager, using GnuPG to encrypt passwords.

.SH COMMANDS

.B init
.RI < gpg-key >
.RS 4
Initialize the password manager using the passed \fIgpg-key\fR as the encryption key. This command must be run first before you start working with LockPassword.
.RE

.B insert
.RB [ "-e, --echo" ]
.RB [ "-c, --copy" ]
.RB [ "-f, --force" ]
.RI < passname >
.RS 4
Add the specified \fIpassname\fR to the password manager. The password will be read interactively using standard input, character display is hidden. The \fB-e, --echo\fR argument enable the show of characters when typing a password; \fB-c, --copy\fR write password to clipboard; \fB-f, --force\fR ignore exist of \fIpassname\fR, overwrites it without prompt.
.RE

.B show
.RB [ "-c, --copy" ]
.RB [ "-C, --no-color ]
.RI < passname >
.RS 4
If no \fIpassname\fR is specified, \fIpassname\fR will assign root of password manager tree. If \fIpassname\fR is directory, list names of passwords inside the at \fIpassname\fR. Otherwise decrypt and print a password of \fIpassname\fR. The \fB-c, --copy\fR write password to clipboard; \fB-C, --no-color\fR cancel colorize the output.
.RE

.B edit
.RI < passname >
.RS 4
Open the specified \fIpassname\fR in a text editor, waiting for changes. The text editor specified in the environment variable EDITOR or \fBvi\fR(1) is used. Temporary file for password editing locate in tmpfs / ramfs (we expect that it to be mounted at /dev/shm).
.RE 

.B generate
.RB [ "-l, --length" =pass-length]
.RB [ "-c, --copy" ]
.RB [ "-f, --force" ]
.RI < passname >
.RS 4
Generate a random password and write it in \fIpassname\fR. The \fB-l, --length = pass-length\fR argument allow you to specify the desired password length. Without this argument, a 14 character password will be generated. \fB-c, --copy\fR write password to clipboard; \fB-f, --force\fR ignore exist of \fIpassname\fR, overwrites it without prompt.
.RE

.B mv
.RB [ "-f, --force" ]
.RI < old-path "> <" new-path >
.RS 4
Move/rename \fIold-path\fR to \fInew-path\fR. \fIold-path\fR must be an exist file, \fInew-path\fR can be a file/directory. The \fB-f, --force\fR argument ignore exist of \fInew-path\fR (if it's a file), overwrites it without prompt.
.RE

.B rm
.RI < passname >
.RS 4
Remove the \fIpassname\fR you specified from the password manager. If the directories where your \fIpassname\fR was nested became empty after deletion, then they are also deleted.
.RE

.B help
.RS 4
Print help information about commands and the application itself.
.RE

.B version
.RS 4
Print information about the version, release date, and license of the application.
.RE

.SH EXAMPLES
Initialize the password manager:
.RS 4
.B $ lpass init joursoir@joursoir.net
.br
LockPassword initialized for joursoir@joursoir.net
.br
.RE

Add password in the password manager:
.RS 4
.B $ lpass insert games/chess/user
.br
Please type your new password: [invisible input]
.br
Please type your new password again: [invisible input]
.br
Password added successfully for games/chess/user
.RE

Print a list of exists password:
.RS 4
.B $ lpass show
.br
Password Manager
.br
\[u251C]\[u2500]\[u2500] banks
.br
\[u2502]   \[u251C]\[u2500]\[u2500] abankpro
.br
\[u2502]   \[u2502]   \[u2514]\[u2500]\[u2500] phone_number
.br
\[u2502]   \[u2514]\[u2500]\[u2500] ubank
.br
\[u2502]       \[u2514]\[u2500]\[u2500] phone_number
.br
\[u2514]\[u2500]\[u2500] games
.br
    \[u2514]\[u2500]\[u2500] chess
.br
        \[u251C]\[u2500]\[u2500] site.com
.br
        \[u2514]\[u2500]\[u2500] user
.RE

Print a list of exists password in some directory:
.RS 4
.B $ lpass show banks
.br
Password Manager/banks
.br
\[u251C]\[u2500]\[u2500] abankpro
.br
\[u2502]   \[u2514]\[u2500]\[u2500] phone_number
.br
\[u2514]\[u2500]\[u2500] ubank
.br
    \[u2514]\[u2500]\[u2500] phone_number
.RE

Show password:
.RS 4
.B $ lpass show games/iko/LordOfNight
.br
helloitismypassword123
.RE

Copy password to clipboard:
.RS 4
.B $ lpass show -c games/iko/LordOfNight
.RE

Generate password:
.RS 4
.B $ lpass generate bank/sbank/phone_number
.br
Generated password: NsNu:+^Re(cshW
.br
Password added successfully for bank/sbank/phone_number
.RE

.SH "COPYING"
Copyright (C) 2020-2025 Aleksandr D. Goncharov (Joursoir)
.br
License: GNU GPL version 3
.br
This is free software: you are free to change and redistribute it. This program comes with ABSOLUTELY NO WARRANTY.


src/lpass.c
/***
	This file is part of LockPassword
	Copyright (C) 2020-2021 Aleksandr D. Goncharov (Joursoir) <chat@joursoir.net>

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
***/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>

#include "constants.h"
#include "exec-cmd.h"
#include "xstd.h"
#include "output.h"
#include "r-lg2.h"

struct cmd_struct {
	const char *cmd;
	int (*func)(int, char **);
};

static struct cmd_struct commands[] = {
	{ "init", cmd_init },
	{ "insert", cmd_insert },
	{ "show", cmd_showtree },
	{ "edit", cmd_edit },
	{ "generate", cmd_generate },
	{ "rm", cmd_remove },
	{ "mv", cmd_move },
	{ "help", cmd_help },
	{ "version", cmd_version },
	{ NULL, NULL }
};

static struct cmd_struct *get_cmd(const char *name)
{
	if(!name)
		return NULL;

	struct cmd_struct *ptr;
	for(ptr = commands; ptr->cmd; ptr++) {
		if(strcmp(name, ptr->cmd) == 0)
			return ptr;
	}
	return NULL;
}

static int goto_maindir()
{
	char *rootdir = xstrcat(getenv("HOME"), "/", LOCKPASS_DIR, NULL);
	int retval = chdir(rootdir);
	if(retval)
	{
		if(errno == ENOENT) {
			retval = mkdir(rootdir, S_IRWXU);
			if(retval && errno != EEXIST) {
				print_error("Error: %s\n", strerror(errno));
			} else {
				retval = chdir(rootdir);
			}
		} else {
			print_error("Error: %s\n", strerror(errno));
		}
	}

	free(rootdir);
	return retval;
}

int main(int argc, char *argv[])
{
	int result;

	if(!isatty(STDIN_FILENO)) {
		print_error("Please, use a terminal to run this application\n");
		return 1;
	}

	if(goto_maindir())
		return 1;

#ifdef LIGBIT
	result = lg2_open_repo(".");
	if (result) {
		print_error("Something went wrong with git\n");
		return 1;
	}
#endif

	struct cmd_struct *ptr = get_cmd(argv[1]);
	result = ptr ? ptr->func(--argc, ++argv) : cmd_help(argc, argv);

#ifdef LIGBIT
	lg2_close_repo();
#endif

	return result;
}